windows bazel build error log:
 ```external/llvm-project/llvm/lib/IR/AutoUpgrade.cpp(3726): fatal error C1061: compiler limit: blocks nested too deeply```
It seems that visual studio does not support variadic templates very well。：）
Should I update the code base? I think it's not my PR problem.

--- a/tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc
+++ b/tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc
@@ -1574,10 +1574,13 @@ class ReduceOnTensorsConversion : public OpConversionPattern<mhlo::ReduceOp> {
 
     SmallVector<Value> inputs, outputs;
     SmallVector<AffineMap, 3> indexing_maps;
-    for (const auto& it : llvm::enumerate(adaptor.inputs())) {
+    for (auto values :
+         llvm::zip(adaptor.inputs(), adaptor.init_values(), op.getResults())) {
       // Check if init_value is constant. If so, inline the value into the
       // region.
-      Value init_value = adaptor.init_values()[it.index()];
+      Value input = std::get<0>(values);
+      Value init_value = std::get<1>(values);
+      Value result = std::get<2>(values);
       Attribute init_const_val = GetInitValueAsConst(init_value);
       if (init_const_val) {
         init_value = rewriter.create<arith::ConstantOp>(
@@ -1586,10 +1589,10 @@ class ReduceOnTensorsConversion : public OpConversionPattern<mhlo::ReduceOp> {
         init_value = rewriter.create<tensor::ExtractOp>(loc, init_value);
       }
 
-      inputs.push_back(it.value());
-      auto result_type = op.getResult(it.index()).getType().cast<ShapedType>();
+      inputs.push_back(input);
+      auto result_type = result.getType().cast<ShapedType>();
       SmallVector<Value, 8> dyn_shape = GetReduceOpInitTensorDynSizes(
-          rewriter, loc, it.value(), result_type, reduction_dims);
+          rewriter, loc, input, result_type, reduction_dims);
       auto init_tensor = GetInitTensor(rewriter, loc, result_type, dyn_shape);
       Value filled_tensor =
           rewriter.create<linalg::FillOp>(loc, init_value, init_tensor)

--- a/tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc
+++ b/tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc
@@ -1563,23 +1563,21 @@ class ReduceOnTensorsConversion : public OpConversionPattern<mhlo::ReduceOp> {
     Location loc = op.getLoc();
 
     int num_inputs = static_cast<int>(adaptor.inputs().size());
-    auto src_type = adaptor.inputs()[0].getType().cast<ShapedType>();
-    int src_rank = src_type.getRank();
-    if (!src_rank) {
+    if (llvm::any_of(adaptor.inputs(), [](Value v) {
+          return !v.getType().cast<ShapedType>().getRank();
+        })) {
       return rewriter.notifyMatchFailure(op, "expects known-rank args");
     }
+    auto src_rank = adaptor.inputs()[0].getType().cast<ShapedType>().getRank();
 
     SmallVector<int64_t, 4> reduction_dims = Extract1DVector(op.dimensions());
 
     SmallVector<Value> inputs, outputs;
     SmallVector<AffineMap, 3> indexing_maps;
-    for (int i = 0; i < num_inputs; ++i) {
-      Value src = adaptor.inputs()[i];
-      if (src.getType() != src_type) return failure();
-
+    for (auto it : llvm::enumerate(adaptor.inputs())) {
       // Check if init_value is constant. If so, inline the value into the
       // region.
-      Value init_value = adaptor.init_values()[i];
+      Value init_value = adaptor.init_values()[it.index()];
       Attribute init_const_val = GetInitValueAsConst(init_value);
       if (init_const_val) {
         init_value = rewriter.create<arith::ConstantOp>(
@@ -1588,10 +1586,10 @@ class ReduceOnTensorsConversion : public OpConversionPattern<mhlo::ReduceOp> {
         init_value = rewriter.create<tensor::ExtractOp>(loc, init_value);
       }
 
-      inputs.push_back(src);
-      auto result_type = op.getResult(i).getType().cast<ShapedType>();
+      inputs.push_back(it.value());
+      auto result_type = op.getResult(it.index()).getType().cast<ShapedType>();
       SmallVector<Value, 8> dyn_shape = GetReduceOpInitTensorDynSizes(
-          rewriter, loc, src, result_type, reduction_dims);
+          rewriter, loc, it.value(), result_type, reduction_dims);
       auto init_tensor = GetInitTensor(rewriter, loc, result_type, dyn_shape);
       Value filled_tensor =
           rewriter.create<linalg::FillOp>(loc, init_value, init_tensor)
@@ -1631,8 +1629,12 @@ class ReduceOnTensorsConversion : public OpConversionPattern<mhlo::ReduceOp> {
     Region& region = linalg_op.region();
     rewriter.inlineRegionBefore(op.body(), region, region.end());
     TypeConverter::SignatureConversion signature_converter(num_inputs * 2);
-    for (int i = 0; i < num_inputs * 2; ++i)
-      signature_converter.addInputs(i, src_type.getElementType());
+
+    // map input and init values's types
+    for (auto it : llvm::enumerate(op.getOperation()->getOperands()))
+      signature_converter.addInputs(
+          it.index(), it.value().getType().cast<ShapedType>().getElementType());
+
     rewriter.applySignatureConversion(&region, signature_converter);
     rewriter.replaceOp(op, linalg_op.getResults());
     return success();

--- a/tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc
+++ b/tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc
@@ -1574,7 +1574,7 @@ class ReduceOnTensorsConversion : public OpConversionPattern<mhlo::ReduceOp> {
 
     SmallVector<Value> inputs, outputs;
     SmallVector<AffineMap, 3> indexing_maps;
-    for (auto it : llvm::enumerate(adaptor.inputs())) {
+    for (const auto& it : llvm::enumerate(adaptor.inputs())) {
       // Check if init_value is constant. If so, inline the value into the
       // region.
       Value init_value = adaptor.init_values()[it.index()];
@@ -1631,9 +1631,10 @@ class ReduceOnTensorsConversion : public OpConversionPattern<mhlo::ReduceOp> {
     TypeConverter::SignatureConversion signature_converter(num_inputs * 2);
 
     // map input and init values's types
-    for (auto it : llvm::enumerate(op.getOperation()->getOperands()))
+    for (const auto& it : llvm::enumerate(op.getOperation()->getOperands())) {
       signature_converter.addInputs(
           it.index(), it.value().getType().cast<ShapedType>().getElementType());
+    }
 
     rewriter.applySignatureConversion(&region, signature_converter);
     rewriter.replaceOp(op, linalg_op.getResults());

--- a/tensorflow/compiler/mlir/hlo/tests/hlo-legalize-to-linalg.mlir
+++ b/tensorflow/compiler/mlir/hlo/tests/hlo-legalize-to-linalg.mlir
@@ -2152,6 +2152,42 @@ func @variadic_reduce(%arg0: tensor<9x2xi32>, %arg1: tensor<9x2xi32>) -> (tensor
 
 // -----
 
+func @variadic_diff_type_reduce(%arg0: tensor<128x10xf32>, %arg1: tensor<128x10xi32>) -> (tensor<128xf32>, tensor<128xi32>) {
+  %cst0 = mhlo.constant dense<1.0> : tensor<f32>
+  %cst1 = mhlo.constant dense<1> : tensor<i32>
+  %res0, %res1 = "mhlo.reduce"(%arg0, %arg1, %cst0, %cst1) ( {
+  ^bb0(%arg7: tensor<f32>, %arg8: tensor<i32>, %arg9: tensor<f32>, %arg10: tensor<i32>):  // no predecessors
+    %0 = "mhlo.compare"(%arg7, %arg9) {comparison_direction = "GE"} : (tensor<f32>, tensor<f32>) -> tensor<i1>
+    %1 = "mhlo.select"(%0, %arg7, %arg9) : (tensor<i1>, tensor<f32>, tensor<f32>) -> tensor<f32>
+    %2 = "mhlo.select"(%0, %arg8, %arg10) : (tensor<i1>, tensor<i32>, tensor<i32>) -> tensor<i32>
+    "mhlo.return"(%1, %2) : (tensor<f32>, tensor<i32>) -> ()
+  }) {dimensions = dense<1> : tensor<1xi64>} : (tensor<128x10xf32>, tensor<128x10xi32>, tensor<f32>, tensor<i32>) ->(tensor<128xf32>, tensor<128xi32>)
+  return %res0, %res1 : tensor<128xf32>, tensor<128xi32>
+}
+// CHECK-DAG:  #[[MAP0:.*]] = affine_map<(d0, d1) -> (d0, d1)>
+// CHECK-DAG:  #[[MAP1:.*]] = affine_map<(d0, d1) -> (d0)>
+// CHECK:      func @variadic_diff_type_reduce
+// CHECK-SAME:   %[[ARG0:[a-zA-Z0-9_]*]]
+// CHECK-SAME:   %[[ARG1:[a-zA-Z0-9_]*]]
+// CHECK:        %[[CST0:.*]] = arith.constant 1.000000e+00 : f32
+// CHECK:        %[[INIT0:.*]] = linalg.init_tensor [128] : tensor<128xf32>
+// CHECK:        %[[FILL0:.*]] = linalg.fill(%[[CST0]], %[[INIT0]])
+// CHECK:        %[[CST1:.*]] = arith.constant 1 : i32
+// CHECK:        %[[INIT1:.*]] = linalg.init_tensor [128] : tensor<128xi32>
+// CHECK:        %[[FILL1:.*]] = linalg.fill(%[[CST1]], %[[INIT1]])
+// CHECK:        %[[RES:.+]]:2 = linalg.generic {
+// CHECK-SAME:     indexing_maps = [#[[MAP0]], #[[MAP0]], #[[MAP1]], #[[MAP1]]]
+// CHECK-SAME:     iterator_types = ["parallel", "reduction"]
+// CHECK-SAME:     ins(%[[ARG0]], %[[ARG1]] : tensor<128x10xf32>, tensor<128x10xi32>)
+// CHECK-SAME:    outs(%[[FILL0]], %[[FILL1]] : tensor<128xf32>, tensor<128xi32>)
+// CHECK-NEXT:   ^bb0(%[[LHS0:.*]]: f32, %[[LHS1:.*]]: i32, %[[RHS0:.*]]: f32, %[[RHS1:.*]]: i32):
+// CHECK-NEXT:      %[[B0:.*]] = arith.cmpf oge, %[[LHS0]], %[[RHS0]] : f32
+// CHECK-NEXT:      %[[RES0:.*]] = select %[[B0]], %[[LHS0]], %[[RHS0]] : f32
+// CHECK-NEXT:      %[[RES1:.*]] = select %[[B0]], %[[LHS1]], %[[RHS1]] : i32
+// CHECK-NEXT:      linalg.yield %[[RES0]], %[[RES1]] : f32, i32
+
+// -----
+
 func @slice_whole_stride(%arg0: tensor<3x4xi32>) -> tensor<1x4xi32> {
   %0 = "mhlo.slice"(%arg0) {
     start_indices = dense<[1, 0]> : tensor<2xi64>,

